### ISSUE TITLE
Qoe Metrics Reporting - Include ReceptionReport@clientID

### ISSUE THREAD

**Issue by dsilhavy** [5G-MAG/rt-5gms-media-stream-handler#70]
## Description
TS 26.512 Section 11.4.3 defines the report format for the Metrics Reporting API. As part of the metrics report the `ReceptionReport@clientID` shall be derived as follows:

> In XML documents representing metrics reports for 3GP-DASH downlink media streaming services, the ReceptionReport@clientID attribute, if present and is available to the Media Session Handler, should be a GPSI value as defined by TS 23.003 [7]. Otherwise, this attribute should be represented by a stable and globally unique string.

## Implementation
We can reuse the existing logic we have for consumption reporting when deriving the `reportingClientId`. For that reason, the code is restructured to create a `reportingClientId` that is provided to the `ConsumptionReportingController` and the `QoEMetricsReportingController`

## Open Questions
Is it valid to use the same ID for `ConsumptionReport.reportingClientId` and `ReceptionReport@clientID`?

**Comment by dsilhavy** [comment]
@rjb1000 Do you have a comment on the open question above regarding: Is it valid to use the same ID for `ConsumptionReport.reportingClientId` and `ReceptionReport@clientID`?

**Comment by rjb1000** [comment]
> Is it valid to use the same ID for _ConsumptionReport.reportingClientId_ and `ReceptionReport@clientID`.

In Rel-18, we are defining a separate media session identifier in TS 26.510 that can be used to correlate consumption reports and QoE metrics reports with 5GMS AS access logs as well as with dynamic policy and Network Assistance invocations. But this is intentionally different from the client identifier, and there is no requirement in TS 26.512 for the same client identifier to be used in consumption reports and QoE metrics reports.

Conversely, though, there is nothing in TS 26.512 to say that the Media Session Handler can't cite the same client identifier in both types of report. If the GPSI is used, these will of course be the same. If it's something generated by the Media Session Handler (e.g. a UUID), it should be fine to use the same value. This would be a natural implementation if the Media Session Handler is producing both kinds of reports from raw data reported by the Media Stream Handler via events and dynamic status information at reference point M7 (see tables 13.2.4-1 and 13.2.6-1 in TS 26.512). If the reports are instead generated in the Media Session Handler, the client identifier(s) would need to be passed back to the Media Stream Handler.

**Comment by dsilhavy** [comment]
Thanks @rjb1000 for the detailed response.

> If the reports are instead generated in the Media Session Handler, the client identifier(s) would need to be passed back to the Media Stream Handler

I assume there is a small mistake here and this should be: _If the reports are instead generated in the **Media Stream Handler**, the client identifier(s) would need to be passed back to the Media Stream Handler_

Let me also quickly explain how we currently implemented this feature because based on your comment I think it does **not** match the expectations. This is something we should probably discuss in one of the next calls as well.

## Current Implementation
We are currently generating the GPSI/UUID directly in the **Media Stream Handler**. The reason for that is that consumption reports and metrics reports are created in the Media Stream Handler as well. That way, we don't need to create and send information from the Media Session Handler that is not directly required by the Media Session Handler itself.

Looking at your comment above, I conclude this behavior is not correct. One shortcoming in our implementation might be that we have a new UUID for each application (because they all include an individual instance of the Media Stream Handler). Generating a UUID in the Media Session Handler enables multiple applications connected to the same Media Session Handler to share a single UUID. All this obviously based on the assumption that by `reporting client` we mean the platform/device itself and not a single application.

## Required changes
If the GPSI/UUID shall be created in the Media Session Handler we need to port this logic from the Media Stream Handler to the Media Session Handler. Once a connection is established, the Media Session Handler can provide the GPSI/UUID to the Media Stream Handler to be included in the consumption and metrics reports.

Note that the reporting procedure is partially event driven. As an example, once a downlink media streaming session has been stopped, a new consumption report is sent. This is handled in the Media Stream Handler that is sending a message to the Media Session Handler with the consumption report. In contrast to consumption reports being triggered via a reporting timer (the timer is managed in the Media Session Handler) there is no direct request by the Media Session Handler before to trigger the reporting. 

````
    fun onPlaybackStateChangedEvent(event: PlaybackStateChangedEvent) {
        if (event.playbackState == PlayerStates.ENDED) {
            triggerConsumptionReport()
        }
    }
````

**Comment by rjb1000** [comment]
(Typo in my comment corrected.)

Let's discuss on Friday's call. The M7 API specification in TS 26.512 clause 13 is not crystal clear about how things should be implemented.

**Comment by dsilhavy** [comment]
Implemented in #67
